tests/testthat/example_analysis.R

```diff
@@ -0,0 +1,85 @@
+# Analysis with all confounders (X, U) being measured. Later on, we assume that
+# U could be missing.
+
+trt_effect <- 1
+data <- simulate_data(N = 10000, alpha_uz = 1, beta_uy = 1,
+                      treatment_effects = trt_effect)
+
+# Implementing PSweight's utility from scratch...
+
+trt_model <- glm(Z ~ X.1 + X.2 + X.3 + U, family = binomial(), data = data)
+e <- predict(trt_model, type = "response")
+weights <- (1 / e) * data$Z + (1 / (1 - e)) * (1 - data$Z)
+Y1_hat <- sum((data$Y * weights)[data$Z == 1]) / sum(weights[data$Z == 1])
+Y0_hat <- sum((data$Y * weights)[data$Z == 0]) / sum(weights[data$Z == 0])
+
+ate_correct_estimate <- Y1_hat - Y0_hat
+
+library(PSweight)
+
+# Using PSweight
+
+ate_correct <- PSweight( # adjusts for confounder U
+  ps.formula = Z ~ X.1 + X.2 + X.3 + U,
+  yname = "Y",
+  family = "binomial",
+  weight = "IPW",
+  ps.method = "glm",
+  data = data
+)
+
+test_that("Adjusting for U yields correct ATE estimates", {
+  expect_equal(ate_correct_estimate, summary(ate_correct)$estimates[1])
+  expect_equal(summary(ate_correct)$estimates[1], trt_effect, tolerance = 0.05)
+})
+
+ate_incorrect <- PSweight( # adjusts for confounder U
+  ps.formula = Z ~ X.1 + X.2 + X.3,
+  yname = "Y",
+  family = "binomial",
+  weight = "IPW",
+  ps.method = "glm",
+  data = data
+)
+
+test_that("Not adjusting for U yields incorrect ATE estimates", {
+  expect_false(summary(ate_incorrect)$estimates[1] == trt_effect)
+})
+
+# What if U were missing?
+# First, consider that we have potential outcomes Y(1) and Y(0)
+# which, by the fundamental problem of causal inference, we don't
+
+run_simulation <- function(seed, c1, c0) {
+
+  data <- simulate_data(N = 1000, alpha_uz = 1, beta_uy = 1,
+                        seed = seed, treatment_effects = trt_effect)
+
+  trt_model <- glm(Z ~ X.1 + X.2 + X.3, family = binomial(), data = data)
+  e <- predict(trt_model, type = "response") # propensity score
+
+  weights <- 1 / ifelse(data$Z, e, 1 - e)
+  Y_sf <- data$Y + (-1)**(data$Z == 1) * abs(data$Z - e) * ifelse(data$Z, c1, c0)
+
+  Y1_sf <- sum((Y_sf * weights)[data$Z == 1]) / sum(weights[data$Z == 1])
+  Y0_sf <- sum((Y_sf * weights)[data$Z == 0]) / sum(weights[data$Z == 0])
+
+  estimated_ate <- Y1_sf - Y0_sf
+
+  return(estimated_ate)
+}
+
+# Because alpha_uz > 0 and beta_uy > 0, treated individuals are more likely to
+# have a better outcome, i.e. both potential outcomes Y(1) and Y(0). Hence, we
+# need c1 > 0 and c0 > 0.
+# Here, c1 = 0.25 and c0 = 0.25 seem to be most valid choices. It is hard to
+# determine the exact numerical value of c1 and c0, but their magnitude can be.
```

Comment:
Does this seem right? @Kuan-Liu The values of `c1 = c0 = 0.25` seem to correspond to our data-generating mechanism, but, in general, the selection of such values seem to be a strong and unavoidable assumption from domain experts